drivers/sensor/adi/adxl372/adxl372_decoder.c

```diff
@@ -0,0 +1,303 @@
+/*
+ * Copyright (c) 2024 Analog Devices Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "adxl372.h"
+
+#ifdef CONFIG_ADXL372_STREAM
+
+/*
+ * Sensor resolution is 100mg/LSB, 12-bit value needs to be right
+ * shifted by 4 or divided by 16. Overall this results in a scale of 160
+ */
+#define SENSOR_SCALING_FACTOR (SENSOR_G / (16 * 1000 / 100))
+#define ADXL372_COMPLEMENT         0xf000
+
+static const uint32_t accel_period_ns[] = {
+	[ADXL372_ODR_400HZ] = UINT32_C(1000000000) / 400,
+	[ADXL372_ODR_800HZ] = UINT32_C(1000000000) / 800,
+	[ADXL372_ODR_1600HZ] = UINT32_C(1000000000) / 1600,
+	[ADXL372_ODR_3200HZ] = UINT32_C(1000000000) / 3200,
+	[ADXL372_ODR_6400HZ] = UINT32_C(1000000000) / 6400,
+};
+
+static inline void adxl372_accel_convert_q31(q31_t *out, const uint8_t *buff)
+{
+	int16_t data_in = ((int16_t)*buff << 4) | (((int16_t)*(buff + 1) & 0xF0) >> 4);
+
+	if (data_in & BIT(11)) {
+		data_in |= ADXL372_COMPLEMENT;
+	}
+
+	int64_t micro_ms2 = data_in * SENSOR_SCALING_FACTOR;
+
+	*out = CLAMP((micro_ms2 + (micro_ms2 % 1000000)), INT32_MIN, INT32_MAX);
```

Comment:
I'm struggling to understand this conversion: I'd expect the output to be q32_t (with its value and shift) and be in m/s^2.

However, I don't see where:
- The shift value in `struct sensor_three_axis_data` is set (zero by default?).
- Where the value is taken from its current unit (micro-m/s^2?) to m/s^2.

Reference:
https://github.com/zephyrproject-rtos/zephyr/blob/69b73ebb0ab9b85fd2f219fa6ca1e135ba58ac88/drivers/sensor/tdk/icm42688/icm42688_decoder.c#L80-L124

May I get some clarification?

Reply:
icm42688 actually also has bugs in its conversion, which is why I created #76968 and @bperseghetti has open #67207 #76861